doc/ircd.8
.\" @(#)ircd.8 7b1 29 Jan 2001
.TH IRCD 8 "ircd-hybrid-7 29 Jan 2001"
.SH NAME
ircd \- The Internet Relay Chat Program Server 
.SH SYNOPSIS
.hy 0
.IP \fBircd\fP
[-tvn] [-f filename] [-d filename] [-k filename] [-h servername]
[-l filename] [-x level]
.SH DESCRIPTION
.LP
\fIircd\fP is the server (daemon) program for the Internet Relay Chat
Program.  The \fIircd\fP is a server in that its function is to "serve"
the client program \fIirc(1)\fP with messages and commands.  All commands
and user messages are passed directly to the \fIircd\fP for processing
and relaying to other ircd sites.
.SH OPTIONS
.TP
.B \-t
Instructs the server to direct debugging output to standard output.
.TP
.B \-f filename
Specifies the ircd.conf file to be used for this ircdaemon. The option
is used to override the default ircd.conf given at compile time.
.TP
.B \-d filename
Specifies the D-line file to be used.  This file is used for both reading
D-lines at startup, and writing to while \fIircd\fP is running.
.TP
.B \-k filename
Specifies the K-line file to be used.  This file is used for both reading
K-lines at startup, and writing to while \fIircd\fP is running.
.TP
.B \-h servername
Specifies the name of the IRC server.  This overrides the setting in
ircd.conf.
.TP
.B \-l filename
Specifies an alternative logfile to be used than that specified in config.h
.TP
.B \-n
Makes \fIircd\fP run in the forground, and, if -x is also specified, output
debugging information to stdout.
.TP
.B \-v
Makes \Fircd\fP print its version, and exit.
.TP
.B \-x <level>
Sets \fPircd\fP's debug level to <level>.
.SH USAGE
If you plan to connect your \fIircd\fP server to an existing Irc-Network,
you will need to alter your local IRC configuration file (typically named
"ircd.conf") so that it will accept and make connections to other \fIircd\fP
servers.  This file contains the hostnames, Network Addresses, and sometimes
passwords for connections to other ircds around the world.  Because 
description of the actual file format of the "ircd.conf" file is beyond the
scope of this document, please refer to the file INSTALL in the IRC source
files documentation directory.
.LP
.SH BOOTING THE SERVER
The \fIircd\fP server can be started as part of the
Unix boot procedure or just by placing the server into Unix Background.
Keep in mind that if it is \fBnot\fP part of your Unix's boot-up procedure 
then you will have to manually start the \fIircd\fP server each time your
Unix is rebooted.  This means if your Unix is prone to crashing
or going for for repairs a lot it would make sense to start the \fIircd\fP
server as part of your UNIX bootup procedure.  
.SH EXAMPLE
.RS
.nf
tolsun% \fBircd\fP
.fi
.RE
.LP
Places \fIircd\fP into Unix background and starts up the server for use.
Note:  You do not have to add the "&" to this command, the program will
automatically detach itself from tty.
.RS
.nf
leguin% \fBircd -n -x 9\fP
.fi
.RE
.LP
Runs ircd in the forground, with maximum debugging information.
.RS
.nf
.SH COPYRIGHT
(c) 1988,1989 University of Oulu, Computing Center, Finland,
.LP
(c) 1988,1989 Department of Information Processing Science,
University of Oulu, Finland
.LP
(c) 1988,1989,1990,1991 Jarkko Oikarinen
.LP
(c) 1997,1998,1999,2000,2001 The IRCD-Hybrid project.
.LP
For full COPYRIGHT see LICENSE file with IRC package.
.LP
.RE
.SH FILES
 "ircd.conf"
.SH "SEE ALSO"
ircd.conf(5)
.SH BUGS
None... ;-) if somebody finds one, please inform author
.SH AUTHOR
irc2.8 and earlier: Jarkko Oikarinen, currently jto@tolsun.oulu.fi.
.LP
ircd-hybrid-7: IRCD-Hybrid Project, ircd-hybrid@the-project.org.
.LP
manual page written by Jeff Trim, jtrim@orion.cair.du.edu,
later modified by jto@tolsun.oulu.fi.
.LP
modified for ircd-hybrid-7 by Edward Brocklesby, ejb@klamath.uucp.leguin.org.uk.



ChangeLog
leeh        2004/01/15 10:13:19 EST	(20040115_3)

  Modified files:
    include              config.h tools.h 
    src                  rserv.c tools.c 
  Log:
  - made dlink_nodes use a balloc
  - add some comments as to which things need to be init'd before what
  
  Revision  Changes    Path
  1.13      +2 -0      rserv/include/config.h
  1.10      +3 -1      rserv/include/tools.h
  1.20      +11 -3     rserv/src/rserv.c
  1.7       +24 -8     rserv/src/tools.c



leeh        2004/01/15 09:50:52 EST	(20040115_2)

  Modified files:
    src                  .cvsignore 
  Added files:
    .                    .cvsignore 
    help                 .cvsignore 
    include              .cvsignore 
  Log:
  - added/updated .cvsignore files
  
  Revision  Changes    Path
  1.1       +4 -0      rserv/.cvsignore (new)
  1.1       +1 -0      rserv/help/.cvsignore (new)
  1.1       +1 -0      rserv/include/.cvsignore (new)
  1.2       +3 -3      rserv/src/.cvsignore



leeh        2004/01/15 09:48:10 EST	(20040115_1)

  Modified files:
    include              client.h config.h tools.h 
    src                  client.c conf.c s_hoststat.c tools.c 
  Log:
  - added an extended hosthash for hoststat service, keeps track of maximum
    host counts etc
  - added get_time() to return time as DD/MM/YY HH:MM TZ
  - oper {}; shouldve been using match(), not strcmp()
  
  Revision  Changes    Path
  1.13      +14 -1     rserv/include/client.h
  1.12      +10 -0     rserv/include/config.h
  1.9       +1 -0      rserv/include/tools.h
  1.16      +57 -0     rserv/src/client.c
  1.14      +4 -4      rserv/src/conf.c
  1.4       +14 -2     rserv/src/s_hoststat.c
  1.6       +14 -0     rserv/src/tools.c



leeh        2004/01/15 09:01:16 EST	(20040115_0)

  Modified files:
    doc                  example.conf 
    include              channel.h 
    src                  channel.c s_alis.c 
  Log:
  - add operbot {}; to example.conf
  - add chmode_to_string_simple() for alis
  
  Revision  Changes    Path
  1.6       +9 -0      rserv/doc/example.conf
  1.8       +1 -0      rserv/include/channel.h
  1.18      +39 -0     rserv/src/channel.c
  1.17      +2 -2      rserv/src/s_alis.c



leeh        2004/01/14 15:59:31 EST	(20040114_0)

  Modified files:
    doc                  example.conf 
    include              config.h newconf.h 
    src                  channel.c newconf.c rserv.c s_alis.c 
                         s_hoststat.c s_operbot.c 
  Log:
  - added service {}; block to conf, to specify a servers n!u@h and gecos
  - added defines for the three services
  - implemented part_service()
  - add operbot {}; block to control which channels operbot joins
  - remove .objoin/.obpart from operbot, theyre not gunna be used
  
  Revision  Changes    Path
  1.5       +18 -0     rserv/doc/example.conf
  1.11      +17 -1     rserv/include/config.h
  1.2       +1 -1      rserv/include/newconf.h
  1.17      +5 -0      rserv/src/channel.c
  1.3       +78 -1     rserv/src/newconf.c
  1.19      +6 -0      rserv/src/rserv.c
  1.16      +4 -0      rserv/src/s_alis.c
  1.3       +4 -0      rserv/src/s_hoststat.c
  1.2       +16 -15    rserv/src/s_operbot.c



leeh        2004/01/13 14:44:16 EST	(20040113_0)

  Modified files:
    include              channel.h 
    src                  c_mode.c cache.c channel.c client.c 
                         tools.c 
  Log:
  - make services rejoining just issue an SJOIN with the client opped
  - added memset()'s where we BlockHeapAlloc()
  - fixed the hosthash
  - made dlink_add() set m->prev = NULL
  
  Revision  Changes    Path
  1.7       +1 -2      rserv/include/channel.h
  1.12      +1 -1      rserv/src/c_mode.c
  1.2       +3 -1      rserv/src/cache.c
  1.16      +4 -6      rserv/src/channel.c
  1.15      +14 -0     rserv/src/client.c
  1.5       +1 -0      rserv/src/tools.c



leeh        2004/01/12 18:40:07 EST	(20040112_0)

  Modified files:
    include              c_init.h channel.h client.h io.h 
    src                  Makefile.in c_mode.c channel.c io.c 
                         rserv.c scommand.c service.c 
  Added files:
    src                  s_operbot.c 
  Log:
  - added an operbot service
  - added storage/functions for services on channels
  - dont introduce services until after EOB
  - made chmode_to_string() give key/limit
  
  Revision  Changes    Path
  1.12      +3 -0      rserv/include/c_init.h
  1.6       +7 -0      rserv/include/channel.h
  1.12      +2 -0      rserv/include/client.h
  1.10      +2 -0      rserv/include/io.h
  1.14      +1 -0      rserv/src/Makefile.in
  1.11      +10 -0     rserv/src/c_mode.c
  1.15      +86 -7     rserv/src/channel.c
  1.19      +0 -2      rserv/src/io.c
  1.18      +1 -0      rserv/src/rserv.c
  1.1       +152 -0    rserv/src/s_operbot.c (new)
  1.14      +2 -0      rserv/src/scommand.c
  1.10      +13 -0     rserv/src/service.c



leeh        2004/01/10 20:27:18 EST	(20040111_2)

  Modified files:
    include              scommand.h stdinc.h 
    src                  c_error.c c_message.c c_mode.c channel.c 
                         client.c io.c scommand.c 
  Log:
  - make parv in the scommand handlers const, otherwise the hooks will have
    issues
  
  Revision  Changes    Path
  1.4       +2 -2      rserv/include/scommand.h
  1.8       +10 -0     rserv/include/stdinc.h
  1.5       +2 -2      rserv/src/c_error.c
  1.13      +13 -7     rserv/src/c_message.c
  1.10      +3 -3      rserv/src/c_mode.c

src/ircd.c
/************************************************************************
 *   IRC - Internet Relay Chat, src/ircd.c
 *   Copyright (C) 1990 Jarkko Oikarinen and
 *                      University of Oulu, Computing Center
 *
 *   This program is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation; either version 1, or (at your option)
 *   any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program; if not, write to the Free Software
 *   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 * $Id$
 */

#include <sys/types.h>
#include <sys/file.h>
#include <sys/stat.h>
#include <sys/socket.h>
#include <sys/wait.h>

#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <errno.h>
#include <time.h>
#include <pwd.h>
#include <fcntl.h>
#include <signal.h>
#include <unistd.h>

#include "setup.h"
#include "config.h"

#ifdef RLIMIT_FD_MAX
#include <sys/time.h>
#include <sys/resource.h>
#endif

#ifdef USE_GETTEXT
#include <libintl.h>
#endif

#ifdef VMS
# include descrip
# include starlet
#endif

#include "tools.h"
#include "ircd.h"
#include "channel.h"
#include "channel_mode.h"
#include "class.h"
#include "client.h"
#include "common.h"
#include "event.h"
#include "fdlist.h"
#include "hash.h"
#include "irc_string.h"
#include "ircdauth.h"
#include "ircd_signal.h"
#include "list.h"
#include "s_gline.h"
#include "motd.h"
#include "ircd_handler.h"
#include "md5.h"
#include "msg.h"         /* msgtab */
#include "hostmask.h"
#include "numeric.h"
#include "parse.h"
#include "res.h"
#include "restart.h"
#include "s_auth.h"
#include "s_bsd.h"
#include "s_conf.h"
#include "s_debug.h"
#include "s_log.h"
#include "s_misc.h"
#include "s_serv.h"      /* try_connections */
#include "s_stats.h"
#include "scache.h"
#include "send.h"
#include "whowas.h"
#include "modules.h"
#include "memory.h"
#include "hook.h"
#include "debug.h"
#include "ircd_getopt.h"
#include "balloc.h"
/*
 * Try and find the correct name to use with getrlimit() for setting the max.
 * number of files allowed to be open by this process.
 */

/* /quote set variables */
struct SetOptions GlobalSetOptions;

/* configuration set from ircd.conf */
struct config_file_entry ConfigFileEntry; 
/* server info set from ircd.conf */
struct server_info ServerInfo;
/* admin info set from ircd.conf */
struct admin_info AdminInfo;

struct  Counter Count;
struct  ServerState_t server_state;

time_t  CurrentTime;            /* GLOBAL - current system timestamp */
int     ServerRunning;          /* GLOBAL - server execution state */
struct Client me;               /* That's me */
struct LocalUser meLocalUser;	/* That's also part of me */

struct Client* GlobalClientList = 0; /* Pointer to beginning of Client list */

struct JupedChannel *JupedChannelList = 0;

/* unknown/client pointer lists */ 
dlink_list unknown_list;        /* unknown clients ON this server only */
dlink_list lclient_list;        /* local clients only ON this server */
dlink_list serv_list;           /* local servers to this server ONLY */
dlink_list global_serv_list;    /* global servers on the network */
dlink_list oper_list;           /* our opers, duplicated in lclient_list */
dlink_list dead_list;           /* clients that have exited, to be freed */

dlink_list lazylink_channels;   /* known about lazylink channels on HUB */
dlink_list lazylink_nicks;	/* known about lazylink nicks on HUB */

int callbacks_called;          /* A measure of server load... */

static unsigned long       initialVMTop = 0;   /* top of virtual memory at init */
static const char * logFileName = LPATH;
static const char * pidFileName = PPATH;

char**  myargv;
int     dorehash   = 0;
int     debuglevel = -1;        /* Server debug level */
char*   debugmode  = "";        /*  -"-    -"-   -"-  */
time_t  nextconnect = 1;        /* time for next try_connections call */

/* Set to zero because it should be initialized later using
 * initialize_server_capabs
 */
int     default_server_capabs = 0x00000000;


/*
 * get_vm_top - get the operating systems notion of the resident set size
 */
static unsigned long get_vm_top(void)
{
  /*
   * NOTE: sbrk is not part of the ANSI C library or the POSIX.1 standard
   * however it seems that everyone defines it. Calling sbrk with a 0
   * argument will return a pointer to the top of the process virtual
   * memory without changing the process size, so this call should be
   * reasonably safe (sbrk returns the new value for the top of memory).
   * This code relies on the notion that the address returned will be an 
   * offset from 0 (NULL), so the result of sbrk is cast to a size_t and 
   * returned. We really shouldn't be using it here but...
   */
#ifndef VMS
  void* vptr = sbrk(0);
  return (unsigned long) vptr;
#else
  return 0;
#endif
}

/*
 * get_maxrss - get the operating systems notion of the resident set size
 */
unsigned long get_maxrss(void)
{
  return get_vm_top() - initialVMTop;
}

/*
 * print_startup - print startup information
 */
static void
print_startup(int pid)
{
  printf("ircd: version %s\n", ircd_version);
  printf("ircd: pid %d\n", pid);
  printf("ircd: running in %s mode from %s\n", !server_state.foreground ? "background"
         : "foreground", ConfigFileEntry.dpath);
}

/*
 * init_sys
 *
 * inputs	- boot_daemon flag
 * output	- none
 * side effects	- if boot_daemon flag is not set, don't daemonize
 */
static void 
init_sys(void)
{
#if defined(RLIMIT_FD_MAX) && !defined(VMS)
  struct rlimit limit;

  if (!getrlimit(RLIMIT_FD_MAX, &limit))
    {

      if (limit.rlim_max < MAXCONNECTIONS)
        {
          fprintf(stderr,"ircd fd table too big\n");
          fprintf(stderr,"Hard Limit: %ld IRC max: %d\n",
                        (long) limit.rlim_max, MAXCONNECTIONS);
          fprintf(stderr,"Fix MAXCONNECTIONS\n");
          exit(-1);
        }

      limit.rlim_cur = limit.rlim_max; /* make soft limit the max */
      if (setrlimit(RLIMIT_FD_MAX, &limit) == -1)
        {
          fprintf(stderr,"error setting max fd's to %ld\n",
                        (long) limit.rlim_cur);
          exit(EXIT_FAILURE);
        }
    }
#endif        /* RLIMIT_FD_MAX */
}

int
make_daemon(void)
{
#ifndef VMS
  int pid;
  
  if((pid = fork()) < 0)
    {
      perror("fork");
      exit(EXIT_FAILURE);
    }
  else if (pid > 0)
    {
      print_startup(pid);
      exit(EXIT_SUCCESS);
    }

  setsid();
  /*  fclose(stdin);
  fclose(stdout);
  fclose(stderr); */
#else
  /* if we get here, assume we've been detached.
     better set a process name. */
  $DESCRIPTOR(myname, "IRCD-HYBRID-7");
  SYS$SETPRN(&myname);
#endif
  return 0;
}

static int printVersion = 0;

struct lgetopt myopts[] = {
  {"dlinefile",  &ConfigFileEntry.dlinefile, 
   STRING, "File to use for dlines.conf"},
  {"configfile", &ConfigFileEntry.configfile, 
   STRING, "File to use for ircd.conf"},
  {"klinefile",  &ConfigFileEntry.klinefile, 
   STRING, "File to use for klines.conf"},
  {"logfile",    &logFileName, 
   STRING, "File to use for ircd.log"},
  {"pidfile",    &pidFileName,
   STRING, "File to use for process ID"},
  {"foreground", &server_state.foreground, 
   YESNO, "Run in foreground (don't detach)"},
  {"version",    &printVersion, 
   YESNO, "Print version and exit"},
#ifdef DEBUGMODE
  {"debug", NULL, 
   ENDEBUG, "Enable debugging for a certain value"},
#endif
  {"help", NULL, USAGE, "Print this text"},
  {NULL, NULL, STRING, NULL},
};

void
set_time(void)
{
  static char to_send[200];
  time_t newtime = time(NULL);
 
  if (newtime == -1)
    {
      ilog(L_ERROR, "Clock Failure (%d)", errno);
      sendto_realops_flags(FLAGS_ALL, L_ALL,
			   "Clock Failure (%d), TS can be corrupted", errno);
      restart("Clock Failure");
    }
  if (newtime < CurrentTime)
    {
      ircsprintf(to_send, "System clock is running backwards - (%lu < %lu)",
		 (unsigned long) newtime,
		 (unsigned long) CurrentTime);

      report_error(L_ALL, to_send, me.name, 0);
      set_back_events(CurrentTime - newtime);
    }
  CurrentTime = newtime;
}

static void
io_loop(void)
{
  int empty_cycles=0, st=0, delay;
  while (ServerRunning)
    {
      /* Run pending events, then get the number of seconds to the next
       * event
       */

      delay = eventNextTime();
      if (delay <= CurrentTime)
	eventRun();
  
      /* Check on the last activity, sleep for up to 1/2s if we are idle... */
      if (callbacks_called > 0)
	empty_cycles = 0;
  
      /* Reset the callback counter... */
      callbacks_called = 0;
      
      if (empty_cycles++ > 10)
	comm_select((st=((empty_cycles-10)*10)>500 ? 500 : st));
      else
	comm_select(0);
  
      /*
       * Check to see whether we have to rehash the configuration ..
       */
      if (dorehash)
	{
	  rehash(&me, &me, 1);
	  dorehash = 0;
	}
    }
}

/*
 * initalialize_global_set_options
 *
 * inputs       - none
 * output       - none
 * side effects - This sets all global set options needed 
 */
static void initialize_global_set_options(void)
{
  memset( &GlobalSetOptions, 0, sizeof(GlobalSetOptions));
  /* memset( &ConfigFileEntry, 0, sizeof(ConfigFileEntry)); */

  GlobalSetOptions.maxclients = MAX_CLIENTS;
  GlobalSetOptions.autoconn = 1;

  GlobalSetOptions.spam_time = MIN_JOIN_LEAVE_TIME;
  GlobalSetOptions.spam_num = MAX_JOIN_LEAVE_COUNT;

  if(ConfigFileEntry.default_floodcount)
    GlobalSetOptions.floodcount = ConfigFileEntry.default_floodcount;
  else
    GlobalSetOptions.floodcount = 10;

  /* memset( &ConfigChannel, 0, sizeof(ConfigChannel)); */

 /* End of global set options */

}

/*
 * initialize_message_files
 *
 * inputs       - none
 * output       - none
 * side effects - Set up all message files needed, motd etc.
 */
static void initialize_message_files(void)
{
  InitMessageFile( HELP_MOTD, HPATH, &ConfigFileEntry.helpfile );
  InitMessageFile( UHELP_MOTD, UHPATH, &ConfigFileEntry.uhelpfile );
  InitMessageFile( USER_MOTD, MPATH, &ConfigFileEntry.motd );
  InitMessageFile( OPER_MOTD, OPATH, &ConfigFileEntry.opermotd );
  InitMessageFile( USER_LINKS, LIPATH, &ConfigFileEntry.linksfile );

  ReadMessageFile( &ConfigFileEntry.helpfile );
  ReadMessageFile( &ConfigFileEntry.uhelpfile );
  ReadMessageFile( &ConfigFileEntry.motd );
  ReadMessageFile( &ConfigFileEntry.opermotd );
  ReadMessageFile( &ConfigFileEntry.linksfile );
}

/*
 * initialize_server_capabs
 *
 * inputs       - none
 * output       - none
 */
static void initialize_server_capabs(void)
{
  /* Default server CAPAB, as defined by CAP_MASK in s_serv.h */
  default_server_capabs = CAP_MASK;

  /* If halfops support is disabled, remove the capab from the list. */
  if (ConfigChannel.use_halfops == 0)
  {
    default_server_capabs &= ~CAP_HOPS;
  }
}


/*
 * write_pidfile
 *
 * inputs       - filename+path of pid file
 * output       - none
 * side effects - write the pid of the ircd to filename
 */
static void write_pidfile(const char *filename)
{
  FBFILE* fd;
  char buff[20];
  if ((fd = fbopen(filename, "w")))
    {
      ircsprintf(buff,"%d\n", (int)getpid());
      if ((fbputs(buff, fd) == -1))
        ilog(L_ERROR,"Error writing to pid file %s (%s)", filename,
		    strerror(errno));
      fbclose(fd);
      return;
    }
  else
    ilog(L_ERROR, "Error opening pid file %s", filename);
}

/*
 * check_pidfile
 *
 * inputs       - filename+path of pid file
 * output       - none
 * side effects - reads pid from pidfile and checks if ircd is in process
 *                list. if it is, gracefully exits
 * -kre
 */
static void check_pidfile(const char *filename)
{
  FBFILE* fd;
  char buff[20];
  pid_t pidfromfile;

  /* Don't do logging here, since we don't have log() initialised */
  if ((fd = fbopen(filename, "r")))
    {
      if (fbgets(buff, 20, fd) == NULL)
	{
	  /*
	    log(L_ERROR, "Error reading from pid file %s (%s)", filename,
	    strerror(errno));
	  */
	}
      else
	{
	  pidfromfile = atoi(buff);
	  if (!kill(pidfromfile, 0))
	    {
	      /* log(L_ERROR, "Server is already running"); */
	      printf("ircd: daemon is already running\n");
	      exit(-1);
	    }
	}
      fbclose(fd);
    }
  else
    {
      /* log(L_ERROR, "Error opening pid file %s", filename); */
    }
}

/*
 * setup_corefile
 *
 * inputs       - nothing
 * output       - nothing
 * side effects - setups corefile to system limits.
 * -kre
 */
static void setup_corefile(void)
{
#ifndef VMS
  struct rlimit rlim; /* resource limits */

  /* Set corefilesize to maximum */
  if (!getrlimit(RLIMIT_CORE, &rlim))
    {
      rlim.rlim_cur = rlim.rlim_max;
      setrlimit(RLIMIT_CORE, &rlim);
    }
#endif
}

/*
 * cleanup_zombies
 * inputs	- nothing
 * output	- nothing
 * side effects - Reaps zombies periodically
 * -AndroSyn
 */
static void cleanup_zombies(void *unused)
{
  int status;
  waitpid(-1, &status, WNOHANG);
}

int main(int argc, char *argv[])
{

  /*
   * save server boot time right away, so getrusage works correctly
   */
  set_time();

  /*
   * Setup corefile size immediately after boot -kre
   */
  setup_corefile();

  /*
   * set initialVMTop before we allocate any memory
   */
  initialVMTop = get_vm_top();

  ServerRunning = 0;

  memset(&me, 0, sizeof(me));
  memset(&meLocalUser, 0, sizeof(meLocalUser));
  me.localClient = &meLocalUser;

  /* Make sure all lists are zeroed */
  memset(&unknown_list, 0, sizeof(unknown_list));
  memset(&lclient_list, 0, sizeof(lclient_list));
  memset(&serv_list, 0, sizeof(serv_list));
  memset(&global_serv_list, 0, sizeof(global_serv_list));
  memset(&oper_list, 0, sizeof(oper_list));
  memset(&lazylink_channels, 0, sizeof(lazylink_channels));
  memset(&lazylink_nicks, 0, sizeof(lazylink_nicks));

  lclient_list.head = lclient_list.tail = NULL;
  oper_list.head = oper_list.tail = NULL;
  serv_list.head = serv_list.tail = NULL;
  global_serv_list.head = global_serv_list.tail = NULL;

  GlobalClientList = &me;       /* Pointer to beginning of Client list */

  memset((void *)&Count, 0, sizeof(Count));
  memset((void *)&server_state, 0, sizeof(server_state));

  Count.server = 1;     /* us */
  memset((void *)&ServerInfo, 0, sizeof(ServerInfo));
  memset((void *)&AdminInfo, 0, sizeof(AdminInfo));

  /* Initialise the channel capability usage counts... */
  init_chcap_usage_counts();

  ConfigFileEntry.dpath = DPATH;
  ConfigFileEntry.configfile = CPATH;   /* Server configuration file */
  ConfigFileEntry.klinefile = KPATH;    /* Server kline file */
  ConfigFileEntry.dlinefile = DLPATH;   /* dline file */
  ConfigFileEntry.glinefile = GPATH;    /* gline log file */

  myargv = argv;
  umask(077);                /* better safe than sorry --SRB */

  parseargs(&argc, &argv, myopts);

  if (printVersion)
  {
    printf("ircd: version %s\n", ircd_version);
    exit(EXIT_SUCCESS);
  }

  if (chdir(ConfigFileEntry.dpath))
  {
    perror("chdir");
    exit(EXIT_FAILURE);
  }

  if (!server_state.foreground)
  {
    make_daemon();
  }
  else
  {
    print_startup(getpid());
  }

  setup_signals();
  /* We need this to initialise the fd array before anything else */
  fdlist_init();
  init_netio();         /* This needs to be setup early ! -- adrian */
  /* Check if there is pidfile and daemon already running */
  check_pidfile(pidFileName);
  /* Init the event subsystem */
  eventInit();
  init_sys();
  if (!server_state.foreground)
  {
    close_all_connections();
  }
  init_log(logFileName);
  initBlockHeap();
  init_dlink_nodes();
  init_resolver();      /* Needs to be setup before the io loop */
  initialize_message_files();
  linebuf_init();       /* set up some linebuf stuff to control paging */
  init_hash();
  id_init();
  clear_scache_hash_table();    /* server cache name table */
  clear_ip_hash_table();        /* client host ip hash table */
  init_host_hash();             /* Host-hashtable. */
  clear_hash_parse();
  init_client();
  initUser();
  init_channels();
  initclass();
  initwhowas();
  init_stats();
  init_hooks();
  load_all_modules(1);
  initServerMask();
  init_auth();                  /* Initialise the auth code */
  read_conf_files(YES);         /* cold start init conf files */
  initialize_server_capabs();   /* Set up default_server_capabs */
  initialize_global_set_options();

  if (ServerInfo.name == NULL)
  {
    fprintf(stderr,
      "ERROR: No server name specified in serverinfo block.\n");
    ilog(L_CRIT,
      "No server name specified in serverinfo block.");
    exit(EXIT_FAILURE);
  }
  /* Can't use strncpy_irc here because we didn't malloc enough... -A1kmm */
  strncpy(me.name, ServerInfo.name, HOSTLEN);

  /* serverinfo{} description must exist.  If not, error out.*/
  if (ServerInfo.description == NULL)
  {
    fprintf(stderr,
      "ERROR: No server description specified in serverinfo block.\n");
    ilog(L_CRIT,
      "ERROR: No server description specified in serverinfo block.");
    exit(EXIT_FAILURE);
  }
  strncpy(me.info, ServerInfo.description, REALLEN);

#ifdef USE_GETTEXT
  /*
   * For 'locale' try (in this order):
   *    Config entry "msglocale" (yyparse() will overwrite LANGUAGE)
   *    Env variable "LANGUAGE"
   *    Default of "" (so don't overwrite LANGUAGE here)
   */

  if (!getenv("LANGUAGE"))
  {
    putenv("LANGUAGE=");
  }

  textdomain("ircd-hybrid");
  bindtextdomain("ircd-hybrid" , MSGPATH);
#endif

#ifdef USE_IAUTH
  iAuth.flags = 0;

  ConnectToIAuth();

  if (iAuth.socket == NOSOCK)
    {
      fprintf(stderr, "Unable to connect to IAuth server\n");
      exit (EXIT_FAILURE);
 }
#endif

  me.fd = -1;
  me.from = &me;
  me.servptr = &me;
  SetMe(&me);
  make_server(&me);
  me.serv->up = me.name;
  me.lasttime = me.since = me.firsttime = CurrentTime;
  add_to_client_hash_table(me.name, &me);
  
  add_server_to_list(&me); /* add ourselves to global_serv_list */
  
  check_class();
  write_pidfile(pidFileName);
  
  ilog(L_NOTICE, "Server Ready");
  
#if 0 && defined(HAVE_LIBCRYPTO)
  eventAdd("cryptlink_regen_key", cryptlink_regen_key, NULL,
	   CRYPTLINK_REGEN_TIME);
#endif

  eventAdd("cleanup_channels", cleanup_channels, NULL, CLEANUP_CHANNELS_TIME);

  eventAdd("cleanup_glines", cleanup_glines, NULL, CLEANUP_GLINES_TIME);

  eventAdd("cleanup_tklines", cleanup_tklines, NULL, CLEANUP_TKLINES_TIME);

  /* We want try_connections to be called as soon as possible now! -- adrian */
  /* No, 'cause after a restart it would cause all sorts of nick collides */
  eventAdd("try_connections", try_connections, NULL, STARTUP_CONNECTIONS_TIME);

  eventAdd("collect_zipstats", collect_zipstats, NULL, ZIPSTATS_TIME);

  /* Setup the timeout check. I'll shift it later :)  -- adrian */
  eventAdd("comm_checktimeouts", comm_checktimeouts, NULL, 1);

  eventAdd("cleanup_zombies", cleanup_zombies, NULL, 30); 
  
#ifdef PACE_CONNECT
 if (ConfigFileEntry.throttle_time > 0)
   eventAdd("flush_expired_ips", flush_expired_ips, NULL, ConfigFileEntry.throttle_time);
 else
   eventAdd("flush_expired_ips", flush_expired_ips, NULL, 300);
#endif

  if(ConfigFileEntry.links_delay > 0)
    eventAdd("write_links_file", write_links_file, NULL, ConfigServerHide.links_delay);

  ServerRunning = 1;
  io_loop();
  return 0;
}



